File: code/012/python/main.py
"""From Python to Go: Python: 012 - User input."""

# Modules
import argparse
from getpass import getpass
from dataclasses import dataclass
import sys


# Classes
@dataclass
class Credentials:
    """Class to store credentials."""
    username: str
    password: str


# Functions
def read_args() -> argparse.Namespace:
    """Helper function to read CLI arguments."""
    parser = argparse.ArgumentParser(description="User input.")
    parser.add_argument("-p", "--path", type=str, help="Path to the input file.")
    return parser.parse_args()


def load_file(path: str) -> str:
    """Function to load a file."""
    with open(path, "r", encoding="utf-8") as file:
        return file.read()


# Main
if __name__ == "__main__":
    # Get arguments
    args = read_args()

    # Load file
    if args.path:
        print(load_file(args.path))

    # Exit if no path provided
    else:
        sys.exit("No path provided.")

    # Get user input
    creds = Credentials(username=input("Username: "), password=getpass("Password: "))
    print(f"{creds=}")

File: code/012/go/main.go
/* From Python to Go: Go(Golang): 012 - User input. */

package main

// Import
import (
	"flag"
	"fmt"
	"os"

	"golang.org/x/term"
)

// types
type CliFlags struct {
	Path string
}

type Credentials struct {
	Username string
	Password string
}

// Functions
func readArgs() CliFlags {
	/* Helper function to read CLI arguments. */

	// Prepare result
	result := CliFlags{}
	flag.StringVar(&result.Path, "p", "", "Path to the input file.")

	// Parse arguments
	flag.Parse()

	// Result
	return result
}

func loadFile(p string) string {
	/* Function to load a file. */
	bs, err := os.ReadFile(p)
	if err != nil {
		os.Exit(2)
	}

	// Result
	return string(bs)
}

func getCreds() Credentials {
	/* Helper function to get credentials */

	// Initialise result
	result := Credentials{}

	// Read Username
	fmt.Print("Username: ")
	_, err := fmt.Scanln(&result.Username)
	if err != nil {
		os.Exit(1)
	}

	// Read password
	fmt.Print("Password: ")
	bytepw, err := term.ReadPassword(int(os.Stdin.Fd()))
	if err != nil {
		os.Exit(1)
	}
	result.Password = string(bytepw)

	// Return result
	return result
}

// Main
func main() {
	/* Main business logic */

	// Get arguments
	arg := readArgs()

	// load file
	if arg.Path != "" {
		fmt.Println(loadFile(arg.Path))

		// Exit if no path provided
	} else {
		os.Exit(3)
	}

	creds := getCreds()
	fmt.Printf("\n%+v\n", creds)
}
